proposals/nnnn-variadic-types.md

```diff
@@ -0,0 +1,239 @@
+
+1. If a generic type that imposes an inferred scalar requirement is applied to a pack element inside a pack expansion, the inferred requirement is a requirement expansion.
+2. If a generic type imposes an inferred requirement expansion, the requirement is expanded for each of the concrete generic arguments.
+3. If a generic type that imposes an inferred requirement expansion is applied to a pack element inside a pack expansion:
+
+    1. The inferred requirement is invalid if it contains multiple pack elements captured by expansions at different depths.
+    2. Otherwise, the nested requirement expansion is semantically equivalent to the innermost requirement expansion.
+  
+  The below code demonstrates each of the above rules:
+
+  ```swift
+  protocol P {
+      associatedtype A
+  }
+  struct ImposeRequirement<T> where T: P {}
+  struct ImposeRepeatedRequirement<each T> where repeat each T: P {}
+  struct ImposeRepeatedSameType<T1: P, each T2> where repeat T1.A == each T2 {}
+  
+  // Infers 'repeat each U: P'
+  func demostrate1<each U>(_: repeat ImposeRequirement<each U>)
+  
+  // Infers 'Int: P, V: P, repeat each U: P'
+  func demostrate2<each U, V>(_: ImposeRepeatedRequirement<Int, V, repeat each U>)
+  
+  // Error. Would attempt to infer 'repeat <U' = each U> repeat <V' = each V> U'.A == V' which is not a supported requirement in the language
+  func demonstrate3a<each U, each V>(_: repeat ImposeRepeatedSameType<each U, repeat each V>))
+  
+  // Infers 'Int: P, repeat each V: P'
+  func demostrate3b<each U, each V>(_: repeat (each U, ImposeRequirementExpansion<Int, repeat each V>))
+```
+
+### Conformances
+
+Variadic structs, classes and actors can conform to protocols. The associated type requirements of the protocol may be fulfilled by a type alias whose underlying type contains pack expansions.
+
+### Type aliases
+
+As with the other variadic types, a variadic type alias either has a generic parameter pack of its own, or can be nested inside of another variadic generic type.
+
+The underlying type of a variadic type alias can reference pack expansion types in the same manner as the type of a stored property. That is, the pack expansions must appear in nested positions, but not at the top level.
+
+```swift
+typealias Element = (repeat each S.Element)
+typealias Callback = (repeat each S) -> ()
+typealias Factory = Other<repeat each S>
+```
+
+Unlike structs and classes, but like other type aliases, variadic type aliases can also be nested inside of generic functions.
+
+### Classes
+
+While there are no restrictions on non-final classes adopting type parameter packs, for the time being the proposal restricts such classes from being the superclass of another class.
```

Comment:
Ought it at least be an error at this stage to declare variadic classes as `open`?

Reply:
We should discuss this during the review. Slava and I figured it might still be useful to be able to declare `open` classes with parameter packs and use them, with the subclassing limitation lifted in a later Swift version (I think the work here is all on the code generation side, right @slavapestov?), because the implementation of this proposal has new runtime support and the `open` vs `public` decision is ABI.

Reply:
Wait, I thought `public` to `open` was a permitted change? I agree with the (similar) rationale below that `final` should not be required. Anyway, yeah, this is appropriate for the review thread.